R/verify.R

```diff
@@ -0,0 +1,52 @@
+#' Reproducibility Check for `@code` Slot in `teal_data`
+#'
+#' Verifies the reproducibility of the `teal_data@env` environment using the `@code` slot.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
+  new_teal_data <- eval_code(teal_data(), x@code)
+
+  reproducible <- all.equal(x@env, new_teal_data@env)
+  if (isTRUE(reproducible)) {
```

Comment:
In case evaluation fails and we end up with a `qenv.error`:
```suggestion
  reproduced <- (inherits(new_teal_data, "teal_data") && isTRUE(all.equal(x@env, new_teal_data@env)))
  if (reproduced) {
```

Reply:
@chlebowa so the result is either `teal_data` or `qenv.error`. If this is a `qenv.error` then `all.equal` brings `FALSE` so I don't think that `inherits(new_teal_data, "teal_data")` is crucial here. From the other hand if `x` is `qenv.error` then this is dispatched to it's own method.

Reply:
Are you sure you won't get an error from `qenv.error@env` first?

Reply:
oh yes, you are right

```r
> x <- teal_data(x1 = iris, code = "stop()")
>   new_teal_data <- eval_code(teal_data(), x@code)
> new_teal_data
<qenv.error:  
 when evaluating qenv code:
stop()>
> all.equal(x@env, new_teal_data@env)
Error in new_teal_data@env : 
  no applicable method for `@` applied to an object of class "qenv.error"
```

merging your suggestion then!